bulk-model-sync-lib/src/commonMain/kotlin/org/modelix/model/sync/bulk/ModelSynchronizer.kt

```diff
@@ -0,0 +1,273 @@
+/*
+ * Copyright (c) 2024.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.modelix.model.sync.bulk
+
+import org.modelix.model.api.ConceptReference
+import org.modelix.model.api.INode
+import org.modelix.model.api.INodeReference
+import org.modelix.model.api.IReferenceLink
+import org.modelix.model.api.IReplaceableNode
+import org.modelix.model.api.IRole
+import org.modelix.model.api.isChildRoleOrdered
+import org.modelix.model.api.remove
+import org.modelix.model.data.NodeData
+
+/**
+ * Similar to [ModelImporter], but the input is two [INode] instances instead of [INode] and [NodeData].
+ *
+ * @param filter determines which nodes need synchronization
+ * @param sourceRoot root of the tree containing the expected nodes
+ * @param targetRoot root of the tree that needs changes
+ * @param nodeAssociation mapping between source and target nodes
```

Comment:
INodeAssociation:associate() is called for each node, that is visited during the sync, in order to avoid looking a node up twice. It is not called for skipped nodes. It ultimately depends on the implementation of INodeAssociation, if additional node associations are maintained.

Reply:
Maybe this would be easier to understand if the kdoc says that this is used for internal optimizations and an appropriate implementation can be passed here?